### ISSUE TITLE
Break CouchApp stuff into a separate module?

### ISSUE THREAD

**Issue by nolanlawson** [pouchdb/express-pouchdb#209]
There's been some discussion in [the CouchDB mailing list](http://couchdb.markmail.org/search/?q=How+do+CouchApps+fit+into+the+CouchDB+story%3F+%28Was%3A+CouchDB+Articles+Pills+and+Tutorials+Ideas%29+order%3Adate-backward#query:How%20do%20CouchApps%20fit%20into%20the%20CouchDB%20story%3F%20%28Was%3A%20CouchDB%20Articles%20Pills%20and%20Tutorials%20Ideas%29%20order%3Adate-backward+page:5+mid:no3jfksdxjcgxz4d+state:results) about where CouchApps actually fit into the CouchDB story, with some indication that maybe they will be broken out into a separate module at some point.

@marten-de-vries I know you put a lot of effort into getting show/update/list/etc. working in pouchdb-server, and I think it definitely makes sense to keep it somewhere, but how easily do you think it could be extracted into an optional component? Do the `mode`s help with that?

Personally I think CouchApps are interesting, but I've just never had any use for them. More and more, web applications are moving toward the "thick-client" model, where the web server is just a simple JSON API and doesn't have much involvement in generating HTML pages. Even where server-side rendering is becoming common (ReactJS, Ember FastBoot), it's coming from a client-oriented design, where the same templating code that's executed on the client is simply pre-executed on the server for a faster initial load.

CouchApps currently have none of this "best of both worlds" approach (isomorphic client/server rendering), and it would be tough to get there unless we managed to have show/list/etc. run within PouchDB on the client, in which case PouchDB becomes a view layer as well as a database. (Kinda weird in my mind. :smiley:)

Basically I agree with what @jchris [said on Twitter](https://twitter.com/jchris/status/595626176950960128). I'd be in favor of breaking CouchApps out into a separate plugin.

@janl @daleharvey thoughts?

**Comment by marten-de-vries** [comment]
> CouchApps currently have none of this "best of both worlds" approach (isomorphic client/server rendering), and it would be tough to get there unless we managed to have show/list/etc. run within PouchDB on the client, in which case PouchDB becomes a view layer as well as a database. (Kinda weird in my mind. :smiley:)

Actually, the plug-ins that power show/update/list etc. in pouchdb-server do also work in the browser on top of PouchDB. Isomorphic client/server rendering is what I had in mind when I build them first. It needs more work (mostly: experimenting with it in a real app setting) though.*

---

The major problem with splitting out the design doc endpoints is that the order of which they're defined in the index.js file matters (Because of function wrapping and the order in which middleware is executed). Everything else is pretty much already separated cleanly. So that's the thing to solve, although I'm not sure how. From a user point of view, the easiest might be to just make the minimumForPouchDB mode the default if CouchDB does make this change.


* Originally I naively hoped I could pouchify existing couchapps with minimal changes, but not being able to intercept browser network requests makes that too hard. Links, xhr and forms are easy. CSS images is harder, but I had it working. DOM modification using JS made it an impossible task, though. :disappointed:

**Comment by nolanlawson** [comment]
Wow, that's actually really cool. :) Isomorphic CouchApps may be a great way to revive them!

**Comment by marten-de-vries** [comment]
I just ran an existing CouchApp on top of PouchDB. Service workers are awesome. [More info](https://gist.github.com/marten-de-vries/3bfaf635f9efdbcf5102).

It's going to take some time though before service workers mature. In the meantime, I have high hopes for a new model of building CouchApps. Basically, you write a JS app as usual (I used React), but use CouchDB to do the prerendering. This does require a few changes to the way you write the JS app (only use the data as available by the current database query), but you get sync for free because of that. And, in the case of React, live updates of the app whenever the database changes. But that approach still needs a bit more work until I can produce a write up.

**Comment by janl** [comment]
Careful down the road of CouchApps, at least at this point.

What I envision is a node server that lives alongside CouchDB that implements all the cool things you could think of, without having to burden the Erlang developers with a feature they don’t care much about.

**Comment by marten-de-vries** [comment]
I vote we continue this in #271 which is very similar. I might add that if we're talking about **just** show/list/update, so not about rewrites/views, splitting them of is probably feasible. For the others, it comes back to the integration problems I've described here and in #271.

Closing, feel free to reopen if you disagree.
